waku/common/enr/typed_record.nim

```diff
@@ -0,0 +1,91 @@
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+
+import
+  std/options,
+  stew/results,
+  eth/keys as eth_keys,
+  libp2p/crypto/crypto as libp2p_crypto
+
+import eth/p2p/discoveryv5/enr except TypedRecord, toTypedRecord
+
+
+## ENR typed record
+
+# Record identity scheme
+
+type RecordId* {.pure.} = enum
+    V4
+
+func toRecordId(id: string): EnrResult[RecordId] =
+  case id:
+  of "v4":
+    ok(RecordId.V4)
+  else:
+    err("unknown identity scheme")
+
+func `$`*(id: RecordId): string =
+  case id:
+  of RecordId.V4: "v4"
+
+
+# Typed record
+
+type TypedRecord* = object
+  raw: Record
+
+proc init(T: type TypedRecord, record: Record): T =
+  TypedRecord(raw: record)
+
+proc tryGet*(record: TypedRecord, field: string, T: type): Option[T] =
+  record.raw.tryGet(field, T)
+
+func toTyped*(record: Record): EnrResult[TypedRecord] =
+  let tr = TypedRecord.init(record)
+
+  # Validate record's identity scheme
+  let idOpt = tr.tryGet("id", string)
+  if idOpt.isNone():
+    return err("missing id scheme field")
+
+  discard ? toRecordId(idOpt.get())
+
+  ok(tr)
+
+
+# Typed record field accessors
+
+func id*(record: TypedRecord): Option[RecordId] =
+  let fieldOpt = record.tryGet("id", string)
+  if fieldOpt.isNone():
+    return none(RecordId)
+
+  let fieldRes = toRecordId(fieldOpt.get())
+  if fieldRes.isErr():
+    return none(RecordId)
+
+  some(fieldRes.value)
+
+func secp256k1*(record: TypedRecord): Option[array[33, byte]] =
+  record.tryGet("secp256k1", array[33, byte])
```

Comment:
probably can be addressed in a follow up pr, can we write a getter for `multiaddrs` too?

Reply:
Coming in a follow-up PR, yes. It will be part of the `waku_enr` module, as it is a non-Ethereum field. It is a Waku-specific field.